python/cuml/metrics/confusion_matrix.py

```diff
@@ -0,0 +1,114 @@
+        or (n_samples, n_outputs)
+        Estimated target values.
+    labels : array-like (device or host) shape = (n_classes,), optional
+        List of labels to index the matrix. This may be used to reorder or
+        select a subset of labels. If None is given, those that appear at least
+        once in y_true or y_pred are used in sorted order.
+    sample_weight : array-like (device or host) shape = (n_samples,), optional
+        Sample weights.
+    normalize : string in [‘true’, ‘pred’, ‘all’]
+        Normalizes confusion matrix over the true (rows), predicted (columns)
+        conditions or all the population. If None, confusion matrix will not be
+        normalized.
+
+    Returns
+    -------
+    C : array-like (device or host) shape = (n_classes, n_classes)
+        Confusion matrix.
+    """
+    y_true, n_rows, n_cols, dtype = \
+        input_to_cuml_array(y_true, check_dtype=[cp.int32, cp.int64])
+    y_true = y_true.to_output('cupy')
+
+    y_pred, _, _, _ = \
+        input_to_cuml_array(y_pred, check_dtype=dtype,
+                            check_rows=n_rows, check_cols=n_cols)
+    y_pred = y_pred.to_output('cupy')
+
+    if labels is None:
+        labels = sorted_unique_labels(y_true, y_pred)
+        n_labels = len(labels)
+    else:
+        labels, n_labels, _, _ = \
+            input_to_cuml_array(labels, check_dtype=dtype, check_cols=1)
+        labels = labels.to_output('cupy')
+    if sample_weight is None:
+        sample_weight = cp.ones(n_rows, dtype=dtype)
+    else:
+        sample_weight, _, _, _ = \
+            input_to_cuml_array(sample_weight,
+                                check_dtype=[cp.float32, cp.float64,
+                                             cp.int32, cp.int64],
+                                check_rows=n_rows, check_cols=n_cols)
+        sample_weight = sample_weight.to_output('cupy')
+
+    if normalize not in ['true', 'pred', 'all', None]:
+        raise ValueError("normalize must be one of {'true', 'pred', "
+                         "'all', None}")
+
+    y_true, _ = make_monotonic(y_true, labels, copy=True)
+    y_pred, _ = make_monotonic(y_pred, labels, copy=True)
+
+    # intersect y_pred, y_true with labels, eliminate items not in labels
+    ind = cp.logical_and(y_pred < n_labels, y_true < n_labels)
+    y_pred = y_pred[ind]
+    y_true = y_true[ind]
+    sample_weight = sample_weight[ind]
+
+    cm = cp.sparse.coo_matrix((sample_weight, (y_true, y_pred)),
+                              shape=(n_labels, n_labels),
+                              dtype=np.float64).toarray()
+
+    with np.errstate(all='ignore'):
+        if normalize == 'true':
+            cm = cp.divide(cm, cm.sum(axis=1, keepdims=True))
+        elif normalize == 'pred':
+            cm = cp.divide(cm, cm.sum(axis=0, keepdims=True))
+        elif normalize == 'all':
+            cm = cp.divide(cm, cm.sum())
+        cm = cp.nan_to_num(cm)
+
+    return cm
```

Comment:
Not necessarily for this PR, but I wonder if we should allow configurable output type in this kind of functions (metrics, preprocessing, etc). CC @JohnZed @cjnolet for their opinion

Reply:
I'm definitely on board with that